### ISSUE TITLE
Pytorch MNIST sample claims to be multi-GPU example, but it's not

### ISSUE THREAD

**Issue by vdabravolski** [awslabs/amazon-sagemaker-examples#1409]
See script here: https://github.com/awslabs/amazon-sagemaker-examples/blob/master/sagemaker-python-sdk/pytorch_mnist/mnist.py#L109

The script is missing starting via multiprocessing.spawn method multiple training processes (1 per GPU). Right now script script is working only for single GPU.

**Comment by metrizable** [comment]
@vdabravolski ,
    
Thank you for using Amazon SageMaker.

I've seen the code you cited and it seems to correctly use `DistributedDataParallel`, as recommended in the PyTorch CUDA docs:

https://pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel

Could you describe in more detail the context of your execution?

**Comment by vdabravolski** [comment]
@metrizable, refer to official [guide for DDP](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel), specifically, this part:
```
In order to spawn up multiple processes per node, you can use either torch.distributed.launch or torch.multiprocessing.spawn
```
Spawning multiple processes per node is missing in SM MIST script.

**Comment by mathephysicist** [comment]
I figured out how to fix this. I will merge a change shortly.

**Comment by hsl89** [comment]
@vdabravolski does the issue still persist?

**Comment by vdabravolski** [comment]
Yes, it's still the same. Nothing changed. The sample as of now, only effective on single-GPU nodes. If you run the same sample on p3.16xlarge you will waste 100/8*7=87,5% of your GPU resources. More importantly, it set erroneous pattern...

If you want to see how multi-GPU/multi-node training should be done, you can refer to [this sample](https://github.com/vdabravolski/mmdetection-sagemaker) and specifically, check this code: https://github.com/vdabravolski/mmdetection-sagemaker/blob/master/container_training/mmdetection_train.py#L199-L234
